### ISSUE TITLE
How to use --auto-iam-authn in my deployment.flile and also getting  Access denied for user <user_iam>@'cloudsqlproxy~ip'

### ISSUE THREAD

**Issue by Levi-coder07** [GoogleCloudPlatform/cloud-sql-proxy#2341]
### Question

How to use --auto-iam-authn correctly and ensure that it works?

### Code

```Go
command:
          - "/cloud-sql-proxy"
          - "--private-ip"
          - "--auto-iam-authn"      
          - "--structured-logs"
          - "--port=3306"
          - "devops-challenge-2024:us-central1:db-bookings"
```

### Additional Details

![image](https://github.com/user-attachments/assets/7b759a9a-5b71-4025-96a9-aea83727bba6)
![image](https://github.com/user-attachments/assets/11ebb9a2-7ec4-4b06-abdc-ad97a50a1b3c)
The proxy works like a charm with the KSA binded with GSA but when i want to enter with the sa-backend-iam with --auto-iam-authn i get the error 
Access denied for user 'sa-backend-iam'@'cloudsqlproxy~179.63.4.243' (using password: YES)

**Comment by jackwotherspoon** [comment]
@Levi-coder07 Thanks for raising an issue on the Cloud SQL Proxy! 

TLDR; Only the SA used to start the Cloud SQL Proxy can login to the database with IAM AuthN

For Cloud SQL IAM Database AuthN to work correctly the IAM principal used to invoke/start the Proxy **must match** that of the IAM principal being used as the IAM database user.

> The proxy works like a charm with the KSA binded with GSA but when i want to enter with the sa-backend-iam with --auto-iam-authn i get the error

If you are starting the Proxy with a Kubernetes service account impersonating an IAM service account, only this SA will be able to login via the Proxy with IAM database authN.

There are two potential solutions to this issue:

1) Start the Proxy as `sa-backend-iam` and then you will be able to successfully login as `sa-backend-iam`
2) Start the Proxy with the KSA using service account account impersonation flag to impersonate `sa-backend-iam`, this will allow you to login as `sa-backend-iam`

How to use the Proxy with SA impersonationn:

https://github.com/GoogleCloudPlatform/cloud-sql-proxy/blob/9cb444cbab9768ab58ae41100b4389b79f5caa97/cmd/root.go#L193-L218

Let me know if you need any further clarification or assistance, happy to help more 😄

**Comment by jackwotherspoon** [comment]
> It is expected behaviour , because my IAM user is not equal to sa which I use to run the proxy ?

@fragile-ds This is exactly correct, it is expected behaviour. 

You will only be able to login using IAM service account principal from `fragile-sql.json` and not your IAM user `xxxx@xxx.com`.

Using service account impersonation as mentioned in my above comment would probably be recommended for your use-case.

**Comment by jackwotherspoon** [comment]
Going to close this out for the time being, if required feel free to re-open 😄

**Comment by fragile-ds** [comment]
@jackwotherspoon U saved my evening , best of us, thx a lot)

**Comment by fragile-ds** [comment]
@jackwotherspoon Hello again,

I followed your instructions step by step, but unfortunately, I still couldn’t make it work. Here’s what I did:

1)Created a new service account with the following roles:
```
roles/cloudsql.admin
roles/cloudsql.instanceUser
roles/iam.serviceAccountTokenCreator
```

2)Enabled IAM authentication on the Cloud SQL database by setting:
`cloudsql.iam_authentication=on`

3)Added the service account to Cloud SQL Users:
`debug-sql@project-dev-428212.iam | IAM (service account)`

4)Ran Cloud SQL Proxy using the service account:

`./cloud-sql-proxy \
  --impersonate-service-account=debug-sql@project-dev-428212.iam.gserviceaccount.com \
  --auto-iam-authn \
  --private-ip \
  --address=0.0.0.0 --port=5432 \
  project-dev-428212:europe-west1:debug-sql
`

5)Tried to connect using the same service account:
`psql -h 127.0.0.1 -U debug-sql@project-dev-428212.iam.gserviceaccount.com -d db
`

Received the following error:
psql: error: connection to server at "127.0.0.1", port 5432 failed: 
FATAL: password authentication failed for user "debug-sql@project-dev-428212.iam.gserviceaccount.com"

**Important Note:**

When I try to connect using my IAM user (e.g., xxxx@xxxxx.com), I receive the error:
FATAL: **Cloud SQL IAM user authentication** failed for user "xxxxx@xxxxx.com"
However, when I attempt to log in using the service account, the database immediately **asks for a password**, as if the service account were treated as a regular PostgreSQL user rather than an IAM principal.

I would appreciate any insight into what might be causing this issue. Thank you in advance
